src/artm/regularizer/net_plsa_phi.cc

```diff
@@ -0,0 +1,166 @@
+      }
+    }
+  }
+
+  for (int token_id = 0; token_id < token_size; ++token_id) {
+    const auto& token_u = p_wt.token(token_id);
+    if (token_u.class_id != config_.class_id())
+      continue;
+
+    auto name_iter = vertex_name2id_.find(token_u.keyword);
+    if (name_iter == vertex_name2id_.end())
+      continue;
+
+    auto edge_iter = edge_weights_.find(name_iter->second);
+    if (edge_iter == edge_weights_.end())
+      continue;
+
+    float D_u = has_weights ? config_.vertex_weight(name_iter->second) : 1.0;
+    for (int topic_id = 0; topic_id < topic_size; ++topic_id) {
+      if (!topics_to_regularize[topic_id])
+        continue;
+
+      float value = 0.0;
+      float p_ut = p_wt.get(token_id, topic_id);
+      for (auto iter = edge_iter->second.begin(); iter != edge_iter->second.end(); ++iter) {
+        auto it = id2vertex_name_.find(iter->first);
+        if (it == id2vertex_name_.end()) {
+          LOG(WARNING) << "Edge links to vertex " << iter->first <<
+                          ", that does not exist in list of vertices, it will be skipped";
+          continue;
+        }
+
+        float p_vt = p_wt.get(p_wt.token_index(
+            ::artm::core::Token(config_.class_id(), it->second)), topic_id);
+        float D_v = has_weights ? config_.vertex_weight(iter->first) : 1.0;
+
+        value += iter->second * (p_vt / (D_u * D_v) - p_ut / (D_u * D_u));
+      }
+      value *= n_t[topic_id] * n_t[topic_id];
+      result->set(token_id, topic_id, value);
+    }
+  }
+
+  return true;
+}
+
+google::protobuf::RepeatedPtrField<std::string> NetPlsaPhi::topics_to_regularize() {
+  return config_.topic_name();
+}
+
+google::protobuf::RepeatedPtrField<std::string> NetPlsaPhi::class_ids_to_regularize() {
+  google::protobuf::RepeatedPtrField<std::string> retval;
+  std::string* ptr = retval.Add();
+  *ptr = config_.class_id();
+  return retval;
+}
+
+void NetPlsaPhi::UpdateNetInfo(const NetPlsaPhiConfig& config) {
+  config_.clear_first_vertex_index();
+  config_.clear_second_vertex_index();
+  config_.clear_edge_weight();
+  config_.clear_vertex_name();
+
+  vertex_name2id_.clear();
+  id2vertex_name_.clear();
+  int num_vertices = config.vertex_name_size();
+  if (num_vertices) {
+    for (int i = 0; i < num_vertices; ++i) {
+      vertex_name2id_[config.vertex_name(i)] = i;
+      id2vertex_name_[i] = config.vertex_name(i);
+    }
+  }
+
+  edge_weights_.clear();
+  int num_edges = config.first_vertex_index_size();
+  if (num_edges) {
+    if (num_edges != config.second_vertex_index_size() || num_edges != config.edge_weight_size()) {
+      BOOST_THROW_EXCEPTION(::artm::core::CorruptedMessageException(
+          "Both vertex indices and value arrays should have the same length"));
+    }
+
+    const auto func = [&](int first_index, int second_index, float value) {  // NOLINT
+      auto iter = edge_weights_.find(first_index);
+      if (iter == edge_weights_.end()) {
```

Comment:
Why not simply write `edge_weights[first_index][second_index] = value`? `map` in c++ creates default values by itself.